### DISCUSSION TITLE
Date Selection Scrip possible?

### DISCUSSION THREAD

**Discussion by stefan-franz** [autokey/autokey#662] [Q&A]
Hello, 
is a command there, that a Date Selection Box appears, i select a date (.e.g next week Tue) and AutoKey put the date in the text after it?

If yes, how is the scrip?

Thanks in Advance, Stefan

**Comment by Elliria** [comment]
There's an AutoKey Calendar dialog that can do that. You can see two example scripts on the [Contributed Scripts 3](https://github.com/autokey/autokey/wiki/Contributed-Scripts-3) page in the AutoKey wiki. In those examples, a dialog is displayed when the date is chosen. The first script demonstrates the default date format provided by the dialog and the second one demonstrates giving you the power to control the format of the date.

If you want just the date to be insert as text in the currently-active window, you would replace this line in either of those two scripts:
```python
dialog.info_dialog(title="The date you chose is:", message=date, width="200")
```
with this line:
```python
keyboard.send_keys(date)
```

**Comment by stefan-franz** [comment]
I tried to use the 3rd script - but there comes a script error - perhaps i need some of the 1st and edit it. But i don't anything about script.

Can anyone please give me the script only to select a date and after choosing, that the date DD-MM-YYYY is paste in the text. 
Thanks in advance

**Reply by Elliria** [reply]
What was the error that you got? I just tested both of those scripts again and they're working at my end, but your error message would hopefully give us a clue as to what went wrong. Anyway, here you go. Here's a script that lets you pick a date from a GUI and then pastes it into the currently-active window in DD-MM-YYYY format, and I'll add it to that wiki page, too, so that it's easier to find in the future:
```python
# Create a variable for the return code and the date and put the chosen date's value into the date variable:
retCode, date = dialog.calendar(title="Choose a date", format="%d-%m-%Y")

# If no date is chosen and the Cancel button is clicked, the Esc key is pressed, or the dialog window is closed:
if retCode:
    # Create a message and display it in a dialog:
    myMessage = "No date was chosen."
    dialog.info_dialog(title="Cancelled", message=myMessage, width="200")
else:
    # Paste the result into the currently-active window:
    keyboard.send_keys(date)
```

**Reply by stefan-franz** [reply]
If i past these script - there happens nothing....is that the whole script?

**Comment by Elliria** [comment]
Yes. Are you pasting it into AutoKey in a script? And are you choosing an Abbreviation of Hotkey to use for triggering the script so that it will run when you're in the window where you'd like the date to be pasted?

**Comment by stefan-franz** [comment]
Yes. The script is called with "abbreviations" and then comes a Script error message in the top right corner of the screen. 
Linux 20.3 Cinnamon
All other scripts like 
```
from datetime import datetime
keyboard.send_keys(datetime.now().strftime('%d.%m.%Y'))
```

are running.

**Comment by Elliria** [comment]
Can you paste the error into here? It might tell us what is going wrong.

**Comment by stefan-franz** [comment]
A black small windows comes on my screen for about 5 seconds - in the top right corner. The message is Autokey - Script Error 
No script of https://github.com/autokey/autokey/wiki/Contributed-Scripts-3 is working - all come with an error.

**Comment by Elliria** [comment]
Can you click on the "Tools" menu when that happens and then choose "View Script Error" from the context menu and then paste what's in that window when it opens? It should display the details of the error.

I'm so sorry that this is happening. Those scripts all work for me, but only you and I have tested them and perhaps there's something about them that isn't good for other systems.

**Comment by stefan-franz** [comment]
The "view script error" is greyed out. It seems something is blocking the script
I use Autokey QT 0.95.10 under Linux Mint 20.3 Cinnamon. 
It seems that the system doesn't allow that the script opens a window (in which i can select the date)

**Comment by Elliria** [comment]
That happens when you haven't had an error. You'll need to cause the error and then click it immediately afterwards. I've also occasionally found that I can't view the script error while the text file that was in the active window is still open, but if I close it, I can. No idea why...

**Reply by josephj11** [reply]
@Elliria I'd like to try to recreate that second case if you can lay out how to reproduce it. I don't think I have seen that before.

**Reply by Elliria** [reply]
I'll let you know the next time I see it. I'm still in the process of going through all my scripts on AutoKey v0.90.4 and in the beta to compare how/whether they work in each and there are several that never worked in the old version of AutoKey. Some of them do that.

**Comment by stefan-franz** [comment]
I found that in Autokey - maybe it helps you

![Bildschirmfoto vom 2022-02-12 08-05-53](https://user-images.githubusercontent.com/96744684/153701102-a1f1f8f4-5e24-44c1-ac61-e79bea3b70d3.jpg)

**Reply by Elliria** [reply]
Interesting. It looks like you don't have KDialog installed. What happens if you use the Gtk version of AutoKey instead of the Qt one?

**Comment by stefan-franz** [comment]
OK - 1st error found. With the Autokey gtk there comes a window - but i need a selection possibility for a date.

My script:
```
# Create a variable for the return code and the date and put the chosen date's value into the date variable:
retCode, date = dialog.calendar(title="Choose a date", format="%d-%m-%Y")

# If no date is chosen and the Cancel button is clicked, the Esc key is pressed, or the dialog window is closed:
if retCode:
    # Create a message and display it in a dialog:
    myMessage = "No date was chosen."
    dialog.info_dialog(title="Cancelled", message=myMessage, width="200")
else:
    # Paste the result into the currently-active window:
    keyboard.send_keys(date)
```

![Bildschirmfoto vom 2022-02-12 19-57-32](https://user-images.githubusercontent.com/96744684/153724551-73af5887-c18f-45aa-a26d-79b1c24668b2.jpg)

**Comment by Elliria** [comment]
Now that's some good progress, because we're at least seeing the script do its thing. I'm not sure why you aren't being offered a selection. Does a little calendar like this one pop up when you run the script?

![Calendar](https://user-images.githubusercontent.com/25792925/153727225-5e135356-153b-4490-8778-c39fc022df6d.png)

If so, you should be able to browse to and click on any date and click the OK button to insert the date into the currently-active window.

**Comment by Elliria** [comment]
It's possible it's the indents that are making it not work, especially if you have other code in the same page. The script you're using is using tabs for the indents. You might want to try replacing each tab with four spaces and see if that makes it happy.

**Comment by stefan-franz** [comment]
I tried Tab instead of 4x spaces. No successful improvement. The same message. 
The date selection don't appears. 
But never mind - the rest works.

**Comment by stefan-franz** [comment]
I tried out some things - and this script appears a  date selection window

```
`# Create a variable for the return code and the date and put the chosen date's value into the date variable:
retCode, date = dialog.calendar(title="Choose a date")
# Use the following line instead if you'd like to control the format of the date:
# retCode, date = dialog.calendar(title="Choose a date", format="%d-%m-%y")

# If no date is chosen and the Cancel button is clicked, the Esc key is pressed, or the dialog window is closed:
if retCode:
    # Create a message and display it in a dialog:
    myMessage = "No date was chosen."
    dialog.info_dialog(title="Cancelled", message=myMessage, width="200")
else:
    # Display the value of the date variable in a dialog:
    dialog.info_dialog(title="The date you chose is:", message=date, width="200")`
```
![Bildschirmfoto vom 2022-02-12 22-21-22](https://user-images.githubusercontent.com/96744684/153728912-258d9295-3d02-43ce-be35-8a4867b625c9.jpg)

![Bildschirmfoto vom 2022-02-12 22-21-34](https://user-images.githubusercontent.com/96744684/153728919-7b73fd90-f8b0-47eb-b1b6-65947e8b0003.jpg)

But not date is inserted in the text editor.

**Comment by Elliria** [comment]
Great progress. Replace this line:
```
    dialog.info_dialog(title="The date you chose is:", message=date, width="200")
```
with this line:
```
    keyboard.send_keys(date)
```
That should paste the date into the currently-active window.

By the way, the only differences between those two scripts are:
- The first one offered you the ability to customize the format of the date (dd-mm-yyyy) and your current one is giving you the default format (yyyy-mm-dd).
- The last line pasted the date into the currently-active window in the previous script and your current one displays it in a dialog.

I'm not sure why the other script didn't work for you.

**Reply by stefan-franz** [reply]
Date is not inserted in the text editor. Nothing happend.

**Reply by Elliria** [reply]
Did you make sure to indent the replacement line the same way the original line was indented?

**Comment by Elliria** [comment]
What happens if you replace this line:
```
retCode, date = dialog.calendar(title="Choose a date")
```
with this line:
```
retCode, date = dialog.calendar(title="Choose a date", format="%d-%m-%Y")
```
I'm hoping that if you can make this change and the one in my previous reply, you'll get a working script that displays the date the way you'd like it to be displayed.

**Reply by stefan-franz** [reply]
Only a window comes: "No date was chosen"

**Reply by Elliria** [reply]
If the indent was the issue and that's now solved, then this should be working. If the indent wasn't the issue and you're still getting "No date was chosen" by replacing just that line, then something about your setup doesn't like ```, format="%d-%m-%Y"``` being added and I'm not sure why. You might want to open an [AutoKey issue](https://github.com/autokey/autokey/issues) about that so that people who are better at coding than I am can take a look at it and hopefully figure it out.

**Reply by josephj11** [reply]
You can try formatting the date in a variable first and then typing it with the API. But, it should work without extra steps like that.

I think there must be something we haven't been able to see yet, because this should just work.
